diff --git a/auto/cc/clang b/auto/cc/clang
index a962ee23..70a6ff15 100644
--- a/auto/cc/clang
+++ b/auto/cc/clang
@@ -89,7 +89,7 @@ if [ "$NGX_SYSTEM" = "Darwin" ]; then
 fi
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
diff --git a/auto/cc/gcc b/auto/cc/gcc
index a5c5c18f..a5ca3369 100644
--- a/auto/cc/gcc
+++ b/auto/cc/gcc
@@ -166,7 +166,7 @@ esac
 
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
diff --git a/auto/cc/icc b/auto/cc/icc
index c47f6e4b..9b35e708 100644
--- a/auto/cc/icc
+++ b/auto/cc/icc
@@ -111,7 +111,7 @@ case "$NGX_ICC_VER" in
 esac
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
