diff -ur nginx-1.7.2/src/core/nginx.h nginx-1.7.2-patched/src/core/nginx.h
--- nginx-1.7.2/src/core/nginx.h	2010-02-12 17:31:01.000000000 +0800
+++ nginx-1.7.2-patched/src/core/nginx.h	2010-03-30 10:52:13.240702627 +0800
@@ -10,7 +10,7 @@
 
 #define nginx_version         1007002
 #define NGINX_VERSION      "1.7.2"
-#define NGINX_VER          "nginx/" NGINX_VERSION
+#define NGINX_VER          "openresty/" NGINX_VERSION ".unknown"
 
 #define NGINX_VAR          "NGINX"
 #define NGX_OLDPID_EXT     ".oldbin"
diff -ur nginx-1.7.2/src/http/ngx_http_header_filter_module.c nginx-1.7.2-patched/src/http/ngx_http_header_filter_module.c
--- nginx-1.7.2/src/http/ngx_http_header_filter_module.c	2010-03-03 23:14:04.000000000 +0800
+++ nginx-1.7.2-patched/src/http/ngx_http_header_filter_module.c	2010-03-30 10:52:53.670909405 +0800
@@ -45,7 +45,7 @@
 };
 
 
-static char ngx_http_server_string[] = "Server: nginx" CRLF;
+static char ngx_http_server_string[] = "Server: openresty" CRLF;
 static char ngx_http_server_full_string[] = "Server: " NGINX_VER CRLF;
 
 
