--- a/src/http/modules/ngx_http_proxy_module.c  Tue Oct 08 16:07:14 2013 +0400
+++ b/src/http/modules/ngx_http_proxy_module.c  Mon Oct 14 10:02:53 2013 +0100
@@ -2796,8 +2796,10 @@ ngx_http_proxy_merge_loc_conf(ngx_conf_t

     if (conf->upstream.upstream || conf->proxy_lengths) {
         clcf = ngx_http_conf_get_module_loc_conf(cf, ngx_http_core_module);
-        if (clcf->handler == NULL && clcf->lmt_excpt) {
-            clcf->handler = ngx_http_proxy_handler;
+        if (clcf->handler == NULL) {
+            if (clcf->lmt_excpt) {
+                clcf->handler = ngx_http_proxy_handler;
+            }
             conf->location = prev->location;
         }
     }
