diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index 80c19656..f79b4a69 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -2153,7 +2156,7 @@ ngx_http_set_virtual_server(ngx_http_request_t *r, ngx_str_t *host)
     ngx_int_t                  rc;
     ngx_http_connection_t     *hc;
     ngx_http_core_loc_conf_t  *clcf;
-    ngx_http_core_srv_conf_t  *cscf;
+    ngx_http_core_srv_conf_t  *cscf = NULL;

 #if (NGX_SUPPRESS_WARN)
     cscf = NULL;
