diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index a5cd44dcc..634dd225b 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -2708,8 +2708,21 @@ ngx_http_finalize_request(ngx_http_request_t *r, ngx_int_t rc)
             return;
         }
 
+#if (NGX_HTTP_V2)
+        if (r->http_version >= NGX_HTTP_VERSION_20
+            && r != r->main
+            && r->post_subrequest
+            && r->lua_subrequest)
+        {
+            /* fall through to normal subrequest finalization */
+        } else {
+            ngx_http_terminate_request(r, rc);
+            return;
+        }
+#else
         ngx_http_terminate_request(r, rc);
         return;
+#endif
     }
 
     if (rc >= NGX_HTTP_SPECIAL_RESPONSE
diff --git a/src/http/ngx_http_request.h b/src/http/ngx_http_request.h
index 6d18cc3c6..df57a021d 100644
--- a/src/http/ngx_http_request.h
+++ b/src/http/ngx_http_request.h
@@ -12,6 +12,8 @@
 #define NGX_HTTP_MAX_URI_CHANGES           10
 #define NGX_HTTP_MAX_SUBREQUESTS           50
 
+#define NGX_HTTP_OPENRESTY_LUA_SUBREQUEST  1
+
 /* must be 2^n */
 #define NGX_HTTP_LC_HEADER_LEN             32
 
@@ -550,6 +552,7 @@ struct ngx_http_request_s {
     unsigned                          internal:1;
     unsigned                          error_page:1;
     unsigned                          filter_finalize:1;
+    unsigned                          lua_subrequest:1;
     unsigned                          post_action:1;
     unsigned                          request_complete:1;
     unsigned                          request_output:1;
